src/main/java/com/wolox/training/security/CustomAuthenticationProvider.java

```diff
@@ -0,0 +1,46 @@
+package com.wolox.training.security;
+
+import com.wolox.training.service.AuthService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+
+@Component
+public class CustomAuthenticationProvider implements AuthenticationProvider {
+
+    @Autowired
+    private AuthService authService;
+
+    @Override
+    public Authentication authenticate(Authentication authentication)
+            throws AuthenticationException {
+
+        String username = authentication.getName();
+        String password = authentication.getCredentials().toString();
+        UserDetails userDetails = authService.loadUserByUsername(username);
+
+        return encoder().matches(password, userDetails.getPassword())
+                ? new UsernamePasswordAuthenticationToken(username, password, new ArrayList<>())
+                : null;
```

Comment:
What happens here if `null` is returned? It throws an authentication error?

Reply:
If null is returned, the request's response is 401 Unauthorized.